Sources/StreamChat/Database/DatabaseContainer.swift

```diff
@@ -50,7 +50,43 @@ class DatabaseContainer: NSPersistentContainer {
         }
         return context
     }()
+    
+    /// An immediately reacting NSManagedObjectContext for the chat state layer.
+    ///
+    /// Chat state layer requires that the context is refreshed when a write happens. Otherwise database observers are too slow to react.
+    ///
+    /// For example, here the state.messages needs to react before loadMessages finishes.
+    /// ```swift
+    /// try await chat.loadMessages()
+    /// let messages = chat.state.messages
+    /// ```
+    private(set) lazy var stateLayerContext: NSManagedObjectContext = {
```

Comment:
Before we were using `writableContext` for reacting to changes immediately in the state layer but that was not a good idea since FRC can update many times within a single write() closure in DatabaseContainer. I was seeing FRC updating itself when the write closure was making a fetch request (flow was fetch message, delete message, fetch message, delete message etc each of the fetch triggered FRC's will/did change).

Only way to get everything working nicely was to manually observe `didSaveObjectsNotification` which means that if the code in the `write` function called `save()` then within the `save()` call, state layer's context merged changes. If we use the `automaticallyMergesChangesFromParent` then the notification handling is triggered with a delay which does not work since it is too late. `write()` exists and `stateLayerContext` still has not refreshed itself which in turn means that FRC has not reacted to the change yet.